File: services/kex/src/telemetry.py
"""Optional OTel tracing to a SELF-HOSTED Arize Phoenix instance.

OFF by default — spans are only recorded when PHOENIX_OTLP_URL is set (e.g.
``http://host.docker.internal:6006/v1/traces``). Traces contain document text,
prompts and extracted knowledge, so this must only ever point at a self-hosted
Phoenix on the local network, never a cloud collector.

Fail-safe by construction: a missing env var, missing opentelemetry packages or
a broken exporter all degrade to a permanent no-op — tracing must never break
or slow the extraction pipeline. Spans carry OpenInference attributes
(``openinference.span.kind``, ``input.value``/``output.value``,
``llm.model_name``) so Phoenix renders proper CHAIN/LLM/TOOL waterfalls.
"""

import os
from contextlib import contextmanager

_SERVICE = os.environ.get("PHOENIX_SERVICE_NAME", "gctrl-kex")

_tracer = None
_init_state = 0  # 0 = not tried, 1 = enabled, -1 = permanently disabled


def _get_tracer():
    global _tracer, _init_state
    if _init_state == 1:
        return _tracer
    if _init_state == -1:
        return None
    url = (os.environ.get("PHOENIX_OTLP_URL") or "").strip()
    if not url:
        _init_state = -1
        return None
    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor

        provider = TracerProvider(
            resource=Resource.create(
                {
                    "service.name": _SERVICE,
                    # Phoenix groups traces into projects by this resource attribute.
                    "openinference.project.name": (os.environ.get("PHOENIX_PROJECT_NAME") or "gctrl").strip() or "gctrl",
                }
            )
        )
        provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=url)))
        trace.set_tracer_provider(provider)
        _tracer = trace.get_tracer(_SERVICE)
        _init_state = 1
        return _tracer
    except Exception:
        _init_state = -1
        return None


class _NoopSpan:
    def set_attribute(self, *_args, **_kwargs):
        pass


_NOOP = _NoopSpan()


def trunc(value, max_len=2000):
    """Cap attribute payloads (documents/prompts) so spans stay lightweight."""
    s = value if isinstance(value, str) else str(value)
    return s[:max_len] + "…" if len(s) > max_len else s


@contextmanager
def span(name, kind="CHAIN", attrs=None):
    """Context-managed span; yields a no-op span object when tracing is off.

    Exceptions propagate unchanged (OTel records them on the span), so the
    pipeline's own try/except degradation paths behave exactly as before.
    """
    tracer = _get_tracer()
    if tracer is None:
        yield _NOOP
        return
    with tracer.start_as_current_span(name) as s:
        try:
            s.set_attribute("openinference.span.kind", kind)
            if attrs:
                for k, v in attrs.items():
                    if v is not None:
                        s.set_attribute(k, trunc(v) if isinstance(v, str) else v)
        except Exception:
            pass
        yield s

File: services/api-rs/src/services/telemetry.rs
//! Optional OTLP export of tracing spans to a SELF-HOSTED Arize Phoenix.
//!
//! OFF by default — spans are only exported when `PHOENIX_OTLP_URL` is set
//! (e.g. `http://host.docker.internal:6006/v1/traces`). When unset, tracing
//! stays exactly as before (plain `fmt` to stdout). A broken exporter degrades
//! to fmt-only and logs the error — telemetry must never take the API down.
//!
//! Spans use OpenInference attributes (`openinference.span.kind`, `input.value`,
//! `gctrl.is_error`) so Phoenix renders them as TOOL/CHAIN nodes, matching the
//! Anvil- and Python-side spans in the same `gctrl` project. Traces contain
//! prompts + knowledge content → self-hosted Phoenix only, never a cloud
//! collector (same stance as the rest of the stack).

use std::sync::OnceLock;

use opentelemetry::trace::TracerProvider as _;
use opentelemetry::KeyValue;
use opentelemetry_otlp::{Protocol, WithExportConfig};
use opentelemetry_sdk::{runtime, trace::TracerProvider, Resource};
use tracing_subscriber::prelude::*;
use tracing_subscriber::EnvFilter;

// Keep the provider alive for the process lifetime; dropping it stops the
// batch exporter background task and silently loses spans.
static PROVIDER: OnceLock<TracerProvider> = OnceLock::new();

/// Initialise tracing. Wires the OTLP→Phoenix layer only when PHOENIX_OTLP_URL
/// is set; otherwise (and on any exporter build error) falls back to the plain
/// fmt subscriber the service always had. Call ONCE at startup.
pub fn init() {
    let url = std::env::var("PHOENIX_OTLP_URL")
        .ok()
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty());

    let Some(url) = url else {
        // Off — unchanged behaviour.
        tracing_subscriber::registry()
            .with(EnvFilter::from_default_env())
            .with(tracing_subscriber::fmt::layer())
            .init();
        return;
    };

    match build_provider(&url) {
        Ok(provider) => {
            let tracer = provider.tracer("gctrl-api");
            let _ = PROVIDER.set(provider);
            tracing_subscriber::registry()
                .with(EnvFilter::from_default_env())
                .with(tracing_subscriber::fmt::layer())
                .with(tracing_opentelemetry::layer().with_tracer(tracer))
                .init();
            tracing::info!("Phoenix tracing enabled → {url}");
        }
        Err(e) => {
            tracing_subscriber::registry()
                .with(EnvFilter::from_default_env())
                .with(tracing_subscriber::fmt::layer())
                .init();
            tracing::error!("Phoenix tracing init failed ({e}) — continuing without export");
        }
    }
}

fn build_provider(url: &str) -> Result<TracerProvider, Box<dyn std::error::Error>> {
    // opentelemetry-otlp 0.27 HTTP uses `with_endpoint` VERBATIM — it does NOT
    // append `/v1/traces` (verified live: a base URL POSTed to `/` → 405). So
    // hand it the FULL traces URL. Our PHOENIX_OTLP_URL already ends in
    // `/v1/traces` (shared with the Python services); ensure it does.
    let mut endpoint = url.trim_end_matches('/').to_string();
    if !endpoint.ends_with("/v1/traces") {
        endpoint.push_str("/v1/traces");
    }

    let exporter = opentelemetry_otlp::SpanExporter::builder()
        .with_http()
        .with_endpoint(endpoint)
        .with_protocol(Protocol::HttpBinary)
        .build()?;

    let project = std::env::var("PHOENIX_PROJECT_NAME")
        .ok()
        .filter(|s| !s.trim().is_empty())
        .unwrap_or_else(|| "gctrl".to_string());

    let provider = TracerProvider::builder()
        .with_batch_exporter(exporter, runtime::Tokio)
        .with_resource(Resource::new(vec![
            KeyValue::new("service.name", "gctrl-api"),
            KeyValue::new("openinference.project.name", project),
        ]))
        .build();
    Ok(provider)
}

/// Cap a span attribute payload (prompts / args / knowledge) so spans stay light.
pub fn trunc(s: &str) -> String {
    const MAX: usize = 2000;
    if s.len() > MAX {
        let mut cut = MAX;
        while !s.is_char_boundary(cut) {
            cut -= 1;
        }
        format!("{}…", &s[..cut])
    } else {
        s.to_string()
    }
}
